extension/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod constants;
+#[cfg(test)]
+mod tests;
+mod v0;
+
+use codec::Encode;
+use constants::*;
+use frame_support::{
+	dispatch::{GetDispatchInfo, PostDispatchInfo},
+	pallet_prelude::*,
+	traits::{Contains, OriginTrait},
+};
+use frame_system::RawOrigin;
+use pallet_contracts::chain_extension::{
+	BufInBufOutState, ChainExtension, Environment, Ext, InitState, RetVal,
+};
+use sp_core::crypto::UncheckedFrom;
+use sp_runtime::{traits::Dispatchable, DispatchError};
+use sp_std::vec::Vec;
+
+type ContractSchedule<T> = <T as pallet_contracts::Config>::Schedule;
+
+pub trait Config:
+	frame_system::Config<RuntimeCall: GetDispatchInfo + Dispatchable<PostInfo = PostDispatchInfo>>
+{
+	/// A query of runtime state.
+	type RuntimeRead: Decode + ReadStateHandler<Self>;
+	/// Allowlisted runtime calls and read state calls.
+	type AllowedApiCalls: Contains<Self::RuntimeCall> + Contains<Self::RuntimeRead>;
+}
+
+/// Trait for handling parameters from the chain extension environment during state read operations.
+pub trait ReadStateHandler<T: Config> {
```

Comment:
Subjective, but the `Handler` suffix doesnt seem necessary to me. Rather than handling a read, isnt it just effectively doing a read.

Would `ReadState::read()` not accomplish the same goal in a much simpler and clearer way?

Reply:
Thanks for this suggestion.